implementations/rust/ockam/ockam_api/src/kafka/portal_listener.rs

```diff
@@ -0,0 +1,72 @@
+use tracing::trace;
+
+use ockam_core::{Address, AllowAll, Any, Route, Routed, Worker};
+use ockam_node::Context;
+
+use crate::kafka::inlet_map::KafkaInletMap;
+use crate::kafka::portal_worker::KafkaPortalWorker;
+use crate::port_range::PortRange;
+
+///First point of ingress of kafka connections, at the first message it spawns new stateful workers
+/// to take care of the connection.
+pub(crate) struct KafkaPortalListener {
+    inlet_map: KafkaInletMap,
+}
+
+#[ockam::worker]
+impl Worker for KafkaPortalListener {
+    type Message = Any;
+    type Context = Context;
+
+    async fn handle_message(
+        &mut self,
+        context: &mut Self::Context,
+        message: Routed<Self::Message>,
+    ) -> ockam::Result<()> {
+        trace!("received first message!");
+        let worker_address = KafkaPortalWorker::start(context, self.inlet_map.clone()).await?;
+
+        //forward to the worker and place its address in the return route
+        let mut message = message.into_local_message();
+
+        message
+            .transport_mut()
+            .onward_route
+            .modify()
+            .replace(worker_address.clone());
+
+        trace!(
+            "forwarding message: onward={:?}; return={:?}; worker={:?}",
+            &message.transport().onward_route,
+            &message.transport().return_route,
+            worker_address
+        );
+
+        context.forward(message).await?;
```

Comment:
The message is converted into a LocalMessage which retains all local metadata, so they are just forwarded "as-is"

Reply:
while in `portal_worker.rs:67` instead, where the ockam message probably won't translate to 1 to 1 to kafka message and is merged and split again, the local information of the last message is copied for every message sent

Reply:
> the local information of the last message is copied for every message sent

"Last message" case would happen if kafka message is in multiple ockam messages, right?
And if it's other way around - we copy the message metadata over multiple outgoing messages, or we repack batched kafka messages into one ockam message?

In Elixir we usually don't forward local metadata. @SanjoDeundiak do we preserve local metadata when forwarding? What's the security implication if we do that?

Reply:
It's an interesting case; as far as I understand, messages that get in the Worker may be split or united (and generally, there is an arbitrary dependency between the number of messages in and out). I'm not sure what is the correct behavior of `LocalInfo` in that case. For now, the only thing we put into `LocalInfo` is SecureChannel metadata (`IdentitiIdentifier`); I think we should be fine for now if we check that the `LocalInfo` of every message coming in is the same. That way, it's OK to clone it to every message that goes out. Not sure it's a future-proof solution, though.

Reply:
> "Last message" case would happen if kafka message is in multiple ockam messages, right?
Yes, we may receive pieces of kafka message 1 byte at a time until it's complete, in this case, we take the local metadata from the ockam message containing the last byte

> And if it's other way around - we copy the message metadata over multiple outgoing messages, or we repack batched kafka messages into one ockam message?
In case multiple kafka messages are received with a single ockam message usually they would be processed and re-forwarded within a single message, but there is no guarantee between ockam messages in and ockam messages out. 
For example, the incoming buffer could be almost full and a single ockam message could bring the last piece of current message and other 3 kafka messages, in this case the current kafka message will be send in several pieces and the most recent 3 kafka message may not end up in the same ockam message.

Reply:
> For now, the only thing we put into LocalInfo is SecureChannel metadata (IdentitiIdentifier)

Yes. And this means that all messages coming from the interceptor are considered coming from a secure channel source. Is that something that we do in other forwarding workers in Rust?
In Elixir I was changing the "source" of the message to "local" when forwarding, because if message is forwarded it's no longer originating from a secure channel.

This behaviour may need some more consistent approach maybe, especially if we add source attributes to the message metadata for ABAC.

Reply:
> Do I understand correctly that there is some mechanism which re-batches messages?

Kafka messages content is currently unchanged, it just gets split and merged back.

> Can you please try to document it somehow, at least in the code comments?

Sure!

Reply:
just pushed the comment commit, marking as resolved